### PR TITLE
Ignore unknown sigalgs and groups in the configuration

### PR DESCRIPTION
**Pull request by t8m**
Related to #20789
    
Unknown signature algorithms and groups in the configuration should not be fatal errors. The handling should be similar to handling of ciphers. I.e., there should be a failure only in case the configuration produces no valid sigalgs or groups.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: test/sslapitest.c [hunk 1]
```diff
@@ -9375,6 +9376,53 @@ static int test_servername(int tst)
...
+        goto end;
+    if (!TEST_size_t_eq(ctx->cert->client_sigalgslen, 2)
+        || !TEST_int_eq(ctx->cert->client_sigalgs[0], TLSEXT_SIGALG_rsa_pkcs1_sha256)
+        || !TEST_int_eq(ctx->cert->client_sigalgs[1], TLSEXT_SIGALG_rsa_pkcs1_sha512))
+        goto end;
+
+#ifndef OPENSSL_NO_EC
+    if (!TEST_int_le(SSL_CTX_set1_groups_list(ctx,
+                                              "nonexistent"),
+                                              0))
```
Comment (beldmit) [inline test/sslapitest.c:9408]: Why is it under `#ifndef OPENSSL_NO_EC`?

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by t8m** [comment]
> I am not sure whether we shouldn't backport it to 3.2.

IMO this would need an OTC exception for backporting to 3.2. Feel free to bring it in an OTC meeting.

**Comment by t8m** [comment]
@beldmit fixup pushed please reconfirm.

**Comment by beldmit** [comment]
** OTC: do we want to backport it to 3.2? As we add support to pluggable signature algorithms in 3.2, this issue would create problems. **

**Comment by t8m** [comment]
Rebased to resolve conflict in CHANGES.md

**Comment by t8m** [comment]
**OTC: This should not be backported to 3.2 branch because this is not a bug fix.**

**Comment by kroeckx** [comment]
I'm against this change. It should return an error if there is some config error, and not do something random different that's totally unexpected. This will be make things very hard to debug what is going on. It's my understanding that an invalid line in a config file is currently just ignored, but it should return an error trying to load the config file.

**Comment by levitte** [comment]
If people are configuring something specific, why shouldn't we raise an error if we can't match what they ask for?  If I was the user, I'd find it very surprising if what I configured was dropped silently...

Viewed from another viewpoint, it may be that we haven't given users good enough tools to express how they want things to work.  Like, what if the user considers some sigalgs mandatory, and others optional based on availability?  That's a viable request in my mind, but I don't think there's a way for them to express that...  and it kinda seems like that's the main question being raised here; are we assuming that the configured sigalgs are mandatory or optional, from the user's perspective?

**Comment by tom-cosgrove-arm** [comment]
As an outsider here, it does seem that if I make a typo and it is silently ignored, that's not good for configuration of a security library...

**Comment by t8m** [comment]
The current code as of all stable versions will silently fall back to the default configuration if you make a typo. That is in case you use the configuration file to configure things. The SSL[_CTX]_set1_client_sigalgs() and SSL[_CTX]_set1_groups_list() will fail if there is a typo in the sigalgs or groups lists but that error is not propagated up when these functions are called during configuration file parsing.

I am going to change this PR to add syntax for optional sigalgs and groups and keep the existing behavior as mandatory.

File: CHANGES.md [hunk 2]
```diff
@@ -28,6 +28,18 @@ OpenSSL 3.3
 
 ### Changes between 3.2 and 3.3 [xx XXX xxxx]
 
+ * Unknown entries in TLS SignatureAlgorithms, ClientSignatureAlgorithms
+   config options and the respective calls to SSL[_CTX]_set1_sigalgs() and
+   SSL[_CTX]_set1_client_sigalgs() that are marked with `?` character are
```
Comment (beldmit) [inline CHANGES.md:33]: s/are marked with/start with/ here and below?

File: test/sslapitest.c [hunk 3]
```diff
@@ -9486,6 +9488,54 @@ static int test_servername(int tst)
...
+    if (!TEST_int_gt(SSL_CTX_set1_groups_list(ctx,
+                                              "P-384:?nonexistent:?P-521"),
+                                              0))
+        goto end;
+    if (!TEST_size_t_eq(ctx->ext.supportedgroups_len, 2)
+        || !TEST_int_eq(ctx->ext.supportedgroups[0], OSSL_TLS_GROUP_ID_secp384r1)
+        || !TEST_int_eq(ctx->ext.supportedgroups[1], OSSL_TLS_GROUP_ID_secp521r1))
+        goto end;
+#endif
+
```
Comment (beldmit) [inline test/sslapitest.c:9532]: Don't we need a negative test here (for 2 unexistent algs)?

**Review by beldmit — APPROVED** [review]
LGTM, have 2 minor amendments/questions

File: test/sslapitest.c (hunk 3 above)
Comment (t8m) [reply test/sslapitest.c:9532]: Could you please be more specific? What should be tested here?

File: CHANGES.md (hunk 2 above)
Comment (t8m) [reply CHANGES.md]: fixed

File: test/sslapitest.c (hunk 3 above)
Comment (beldmit) [reply test/sslapitest.c:9532]: Sure. If we specify only non-existent algorithms, we should get an error. 

Smth like 
```
if (TEST_int_gt(SSL_CTX_set1_sigalgs_list(ctx,
                                               "?nonexistent1:?nonexistent1:?nonexistent2"),
                                               0))
```

**Review by beldmit — APPROVED** [review]
My approval persists. Many thanks for amendments!

File: CHANGES.md [hunk 4]
```diff
@@ -28,6 +28,19 @@ OpenSSL 3.3
 
 ### Changes between 3.2 and 3.3 [xx XXX xxxx]
 
+ * Unknown entries in TLS SignatureAlgorithms, ClientSignatureAlgorithms
+   config options and the respective calls to SSL[_CTX]_set1_sigalgs() and
+   SSL[_CTX]_set1_client_sigalgs() that start with `?` character are
+   ignored and the configuration will still be used.
```
Comment (kroeckx) [inline CHANGES.md:34]: Is that ignored if it's unknown, but used if it's known?

Comment (t8m) [reply CHANGES.md:34]: I meant that the whole configuration will be still used even if the list items marked with `?` are unknown.

**Comment by beldmit** [comment]
Can we please get an approval from @openssl/commiters?

**Review by romen — APPROVED** [review]
I have been looking at this for the past couple weeks and it looks good for me.

**Comment by beldmit** [comment]
@t8m could you please rebase it to eliminate the CHANGES.md conflict?

**Comment by beldmit** [comment]
Merged with fixing the conflict in CHANGES.md
Many thanks for this work and review!

**Pull request closed without merging**